File: backend/src/main/java/com/questbase/backend/relationship/questnpc/QuestNpc.java
package com.questbase.backend.relationship.questnpc;

import com.questbase.backend.npc.Npc;
import com.questbase.backend.quest.Quest;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(
    name = "quest_npcs",
    uniqueConstraints = {
        @UniqueConstraint(columnNames = {"quest_id", "npc_id"})
    }
)
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class QuestNpc {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "quest_id", nullable = false)
    private Quest quest;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "npc_id", nullable = false)
    private Npc npc;

    private String role;

    @Column(columnDefinition = "TEXT")
    private String notes;
}

File: frontend/src/types/api/questnpc.ts
import { Npc } from "./npc"
import { Quest } from "./quest"

export interface QuestNpc {
  id: number,
  questId: number,
  npcId: number,
  role: string,
  notes: string
  npc: Npc
}

export interface NpcQuest {
  id: number,
  questId: number,
  npcId: number,
  role: string,
  notes: string
  quest: Quest
}

export interface CreateQuestNpcRequest {
  questId: number,
  npcId: number,
  role: string,
  notes: string
}

export interface UpdateQuestNpcRequest {
  role: string,
  notes: string
}